mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(const func::FuncOp &func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](Operation *op) {
+            if (op->getName().getStringRef() == "quantum.device") {
+                count++;
+            }
+        });
+        return count;
+    } // countTapes()
+
+    void collectOperationsForEachTape(
+        const func::FuncOp &func,
+        SmallVector<std::shared_ptr<SmallVector<Operation *>>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        for (size_t i = 0; i < OpsEachTape.size(); i++) {
+            OpsEachTape[i] = std::make_shared<SmallVector<Operation *>>();
+        }
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int cur_tape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (op->getName().getStringRef() == "quantum.device") {
+                cur_tape++;
+            }
+            OpsEachTape[cur_tape]->push_back(op);
+            if ((op->getName().getStringRef() == "quantum.device_release") &&
+                cur_tape == OpsEachTape.size() - 2) {
+                // reached post processing
+                cur_tape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    void collectNecessaryValuesFromEarlierTapes(const SmallVector<Operation *> &TapeOps,
```

Comment:
Is it not the collection of values from the preprocessing? I am not sure why tapes would share some values?

Reply:
In principle, a later tape could use an earlier tape's values (e.g. use the returned expval as rotation angles, etc). Although it is quite hard to write this in frontend's PL syntax, it is not impossible, e.g. with calling PL's `tape` object's constructor directly.